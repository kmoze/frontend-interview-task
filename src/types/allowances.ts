export type Allowance = {
  id: number
  name: string
  renewal: 'week' | 'month' | 'year'
  currency: string
  amount: string
  spent: string
  active: boolean
  type: 'expense' | 'card'
}
