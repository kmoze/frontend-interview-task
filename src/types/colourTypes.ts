import { COLOUR_OPTIONS } from '@/constants/colourOptions'

export type ColourOption = (typeof COLOUR_OPTIONS)[keyof typeof COLOUR_OPTIONS]
