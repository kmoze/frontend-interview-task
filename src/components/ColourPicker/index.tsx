import React, { useState } from 'react'
import { ColourOption } from '@/types/colourTypes'
import { Palette } from 'lucide-react'

type ColourPickerProps = {
  colours: { [key: string]: ColourOption }
  selectedColour: ColourOption
  onColourSelect: (colour: ColourOption) => void
}

const ColourPicker = ({
  colours,
  selectedColour,
  onColourSelect,
}: ColourPickerProps) => {
  const [isOpen, setIsOpen] = useState<boolean>(false)

  return (
    <div className="fixed bottom-6 right-6 flex items-end gap-2 z-10">
      <div
        className={`flex gap-2 transition-all duration-300 ease-in-out ${
          isOpen
            ? 'opacity-100 translate-y-0'
            : 'opacity-0 translate-y-4 pointer-events-none'
        }`}
      >
        {Object.entries(colours).map(([name, colour]) => (
          <button
            key={name}
            onClick={() => {
              onColourSelect(colour)
              setIsOpen(false)
            }}
            className={`w-8 h-8 rounded-full border-2 ${
              selectedColour === colour ? 'border-black' : 'border-gray-100'
            } hover:scale-110 transition-transform hover:-translate-y-0.5 hover:shadow-lg hover:border-gray-500
          duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-gray-500`}
            style={{ backgroundColor: colour }}
            aria-label={`Select ${name} theme`}
          />
        ))}
      </div>
      <button
        onClick={() => setIsOpen(!isOpen)}
        className={`p-2 rounded-full bg-white shadow-lg transition-all duration-200
        ${isOpen ? 'bg-gray-100' : ''}`}
        aria-label="Toggle colour picker"
      >
        <Palette
          size={24}
          className={`transition-transform duration-300 text-[${selectedColour}] ${isOpen ? 'rotate-90' : ''}`}
        />
      </button>
    </div>
  )
}

export default ColourPicker
