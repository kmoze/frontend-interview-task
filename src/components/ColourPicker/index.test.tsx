import { render, fireEvent } from '@testing-library/react'
import ColourPicker from '@/components/ColourPicker'
import '@testing-library/jest-dom'
import { ColourOption } from '@/types/colourTypes'

describe('ColourPicker Component', () => {
  const mockColours: { [key: string]: ColourOption } = {
    cherryRed: '#74070E',
    butterYellow: '#FFEDA8',
    auraIndigo: '#B0A6DE',
  }

  const mockOnColourSelect = jest.fn()

  it('renders all colour buttons', () => {
    const { getByLabelText } = render(
      <ColourPicker
        colours={mockColours}
        selectedColour="#B0A6DE"
        onColourSelect={mockOnColourSelect}
      />
    )

    expect(getByLabelText('Select cherryRed theme')).toBeInTheDocument()
    expect(getByLabelText('Select butterYellow theme')).toBeInTheDocument()
    expect(getByLabelText('Select auraIndigo theme')).toBeInTheDocument()
  })

  it('applies correct border styling for the selected colour', () => {
    const { getByLabelText } = render(
      <ColourPicker
        colours={mockColours}
        selectedColour="#B0A6DE"
        onColourSelect={mockOnColourSelect}
      />
    )

    const selectedButton = getByLabelText('Select auraIndigo theme')
    expect(selectedButton).toHaveClass('border-black')

    const otherButton = getByLabelText('Select cherryRed theme')
    expect(otherButton).toHaveClass('border-gray-100')
  })

  it('calls onColourSelect with correct colour when clicked', () => {
    const { getByLabelText } = render(
      <ColourPicker
        colours={mockColours}
        selectedColour="#B0A6DE"
        onColourSelect={mockOnColourSelect}
      />
    )

    const indigoButton = getByLabelText('Select auraIndigo theme')
    fireEvent.click(indigoButton)

    expect(mockOnColourSelect).toHaveBeenCalledTimes(1)
    expect(mockOnColourSelect).toHaveBeenCalledWith('#B0A6DE')
  })

  it('toggles colour picker visibility when palette icon is clicked', () => {
    const { getByLabelText, getByRole } = render(
      <ColourPicker
        colours={mockColours}
        selectedColour="#B0A6DE"
        onColourSelect={mockOnColourSelect}
      />
    )

    const toggleButton = getByLabelText('Toggle colour picker')
    const colourPickerContainer = getByRole('button', {
      name: 'Select cherryRed theme',
    }).parentElement

    expect(colourPickerContainer).toHaveClass('opacity-0')
    expect(colourPickerContainer).toHaveClass('pointer-events-none')

    fireEvent.click(toggleButton)
    expect(colourPickerContainer).toHaveClass('opacity-100')
    expect(colourPickerContainer).not.toHaveClass('pointer-events-none')

    fireEvent.click(toggleButton)
    expect(colourPickerContainer).toHaveClass('opacity-0')
    expect(colourPickerContainer).toHaveClass('pointer-events-none')
  })

  it('hides colour picker after selecting a colour', () => {
    const { getByLabelText, getByRole } = render(
      <ColourPicker
        colours={mockColours}
        selectedColour="#B0A6DE"
        onColourSelect={mockOnColourSelect}
      />
    )

    const toggleButton = getByLabelText('Toggle colour picker')
    const colourPickerContainer = getByRole('button', {
      name: 'Select cherryRed theme',
    }).parentElement

    fireEvent.click(toggleButton)
    expect(colourPickerContainer).toHaveClass('opacity-100')

    const cherryRedButton = getByLabelText('Select cherryRed theme')
    fireEvent.click(cherryRedButton)

    expect(colourPickerContainer).toHaveClass('opacity-0')
    expect(colourPickerContainer).toHaveClass('pointer-events-none')
  })

  it('renders with a snapshot', () => {
    const { container } = render(
      <ColourPicker
        colours={mockColours}
        selectedColour="#B0A6DE"
        onColourSelect={mockOnColourSelect}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
