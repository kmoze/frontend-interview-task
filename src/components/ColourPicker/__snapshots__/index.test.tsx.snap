// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColourPicker Component renders with a snapshot 1`] = `
<div>
  <div
    class="fixed bottom-6 right-6 flex items-end gap-2 z-10"
  >
    <div
      class="flex gap-2 transition-all duration-300 ease-in-out opacity-0 translate-y-4 pointer-events-none"
    >
      <button
        aria-label="Select cherryRed theme"
        class="w-8 h-8 rounded-full border-2 border-gray-100 hover:scale-110 transition-transform hover:-translate-y-0.5 hover:shadow-lg hover:border-gray-500
          duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-gray-500"
        style="background-color: rgb(116, 7, 14);"
      />
      <button
        aria-label="Select butterYellow theme"
        class="w-8 h-8 rounded-full border-2 border-gray-100 hover:scale-110 transition-transform hover:-translate-y-0.5 hover:shadow-lg hover:border-gray-500
          duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-gray-500"
        style="background-color: rgb(255, 237, 168);"
      />
      <button
        aria-label="Select auraIndigo theme"
        class="w-8 h-8 rounded-full border-2 border-black hover:scale-110 transition-transform hover:-translate-y-0.5 hover:shadow-lg hover:border-gray-500
          duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-gray-500"
        style="background-color: rgb(176, 166, 222);"
      />
    </div>
    <button
      aria-label="Toggle colour picker"
      class="p-2 rounded-full bg-white shadow-lg transition-all duration-200
        "
    >
      <svgr-url
        class="w-6 h-6 transition-transform duration-300 "
        style="stroke: #B0A6DE;"
      />
    </button>
  </div>
</div>
`;
